asset-transfer-ledger-queries/chaincode-javascript/lib/asset_transfer_ledger_chaincode.js

```diff
@@ -0,0 +1,425 @@
+    } catch (err) {
+      jsonResp = {};
+      jsonResp.error = 'Failed to decode JSON of: ' + id;
+      throw new Error(jsonResp);
+    }
+    await ctx.stub.deleteState(id); //remove the asset from chaincode state
+
+    // delete the index
+    let indexName = 'color~name';
+    let colorNameIndexKey = ctx.stub.createCompositeKey(indexName, [assetJSON.color, assetJSON.ID]);
+    if (!colorNameIndexKey) {
+      throw new Error(' Failed to create the createCompositeKey');
+    }
+    //  Delete index entry to state.
+    await ctx.stub.deleteState(colorNameIndexKey);
+  }
+
+  // TransferAsset transfers a asset by setting a new owner name on the asset
+  async transferAsset(ctx, assetName, newOwner) {
+
+    let assetAsBytes = await ctx.stub.getState(assetName);
+    if (!assetAsBytes || !assetAsBytes.toString()) {
+      throw new Error(`Asset ${assetName} does not exist`);
+    }
+    let assetToTransfer = {};
+    try {
+      assetToTransfer = JSON.parse(assetAsBytes.toString()); //unmarshal
+    } catch (err) {
+      let jsonResp = {};
+      jsonResp.error = 'Failed to decode JSON of: ' + assetName;
+      throw new Error(jsonResp);
+    }
+    assetToTransfer.owner = newOwner; //change the owner
+
+    let assetJSONasBytes = Buffer.from(JSON.stringify(assetToTransfer));
+    await ctx.stub.putState(assetName, assetJSONasBytes); //rewrite the asset
+  }
+
+// GetAssetsByRange performs a range query based on the start and end keys provided.
+// Read-only function results are not typically submitted to ordering. If the read-only
+// results are submitted to ordering, or if the query is used in an update transaction
+// and submitted to ordering, then the committing peers will re-execute to guarantee that
+// result sets are stable between endorsement time and commit time. The transaction is
+// invalidated by the committing peers if the result set has changed between endorsement
+// time and commit time.
+// Therefore, range queries are a safe option for performing update transactions based on query results.
+  async getAssetsByRange(ctx, startKey, endKey) {
+
+    let resultsIterator = await ctx.stub.getStateByRange(startKey, endKey);
+    let results = await this.getAllResults(resultsIterator, false);
+
+    return JSON.stringify(results);
+  }
+
+  // TransferAssetBasedOnColor will transfer assets of a given color to a certain new owner.
+  // Uses a GetStateByPartialCompositeKey (range query) against color~name 'index'.
+  // Committing peers will re-execute range queries to guarantee that result sets are stable
+  // between endorsement time and commit time. The transaction is invalidated by the
+  // committing peers if the result set has changed between endorsement time and commit time.
+  // Therefore, range queries are a safe option for performing update transactions based on query results.
+  // Example: GetStateByPartialCompositeKey/RangeQuery
+  async transferAssetsBasedOnColor(ctx, color, newOwner) {
```

Comment:
Brett is changing this to "transferAssetByColor" in the go chaincode PR, might as well do the same here.